### PR TITLE
Feature/soul bound tokens

### PR DESCRIPTION
**Pull request by Khairul-Blockcode**
more tests added for SBT

### CODE REVIEW

File: contracts/factoryContracts/SBTFactory.sol [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+
+    mapping(uint256 => UserInfo) internal _users;
+    mapping(uint256 => BurnAuth) internal _auth;
+    mapping(uint256 => string) private token_Uri;
+
+
+    event ExpiryExtended(uint256 newExpiration, uint256 _tokenId);
+    event UriChanged(uint256 tokenId, string newuri);
+
+	event Whitelisted(address user, bool value);
```
Comment (dnzdlklc) [inline contracts/factoryContracts/SBTFactory.sol:35]: Indentation

File: contracts/factoryContracts/SBTFactory.sol [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+        emit Issued(msg.sender, _recipient, id, _auth[id]);
+    }
+
+    /* @dev  MRHB or the whitelisted member can call this function to mint/issue 5 SBT at at time*/
+    function issueMany(
+        address[] memory _recipient,
+        string[] memory _uri,
+        uint256[] memory _expires
+    ) external {
+        require(_recipient.length <= maxMintLimit, "SBT: Number of reciepient exceed the max mint limit");
```
Comment (dnzdlklc) [inline contracts/factoryContracts/SBTFactory.sol]: change error message to "SBT: Exceeds Max Mint Limit Per Call"

File: contracts/factoryContracts/SBTFactory.sol [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+    /* @dev  sets user of SB tokens*/
+    function _setUser(
+        uint256 tokenId,
+        address user,
+        uint256 expires
+    ) internal {
+        UserInfo storage info = _users[tokenId];
+        info.user = user;
+        info.expires = expires;
+        // emit UpdateUser(tokenId, user, expires);
```
Comment (dnzdlklc) [inline contracts/factoryContracts/SBTFactory.sol]: Uncomment this event please

File: contracts/factoryContracts/SBTFactory.sol [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+        require(newExpiration > block.timestamp, " Not valid time");
+        require(_exists(_tokenId), "ERC721Metadata: URI query for nonexistent token");
+        _users[_tokenId].expires = newExpiration;
+
+        emit ExpiryExtended(newExpiration, _tokenId);
+    }
+
+    function setWhitelistEnabled(bool _state) public onlyOwner {
+        whitelistEnabled = _state;
+		emit WhiteListEnabled(_state);
```
Comment (dnzdlklc) [inline contracts/factoryContracts/SBTFactory.sol:138]: Indentation

File: contracts/factoryContracts/SBTFactory.sol [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+    }
+
+    function setWhitelist(address[] calldata newAddresses) public onlyOwner {
+        // At least one royaltyReceiver is required.
+        require(newAddresses.length > 0, "No user details provided");
+        // Check on the maximum size over which the for loop will run over.
+        require(newAddresses.length < 2000, "Too many userss to whitelist");
```
Comment (dnzdlklc) [inline contracts/factoryContracts/SBTFactory.sol]: typo single "s"

File: contracts/factoryContracts/SBTFactory.sol [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+        for (uint256 i = 0; i < newAddresses.length; i++) {
+            require(mapSize < 2000, "Maximum Users already whitelisted");
+            whitelist[newAddresses[i]] = true;
+            mapSize++;
+			emit Whitelisted(newAddresses[i],true);
```
Comment (dnzdlklc) [inline contracts/factoryContracts/SBTFactory.sol:160]: indendation

File: contracts/factoryContracts/SBTFactory.sol [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+    }
+
+    function removeWhitelist(address[] calldata currentAddresses) public onlyOwner {
+        // At least one royaltyReceiver is required.
+        require(currentAddresses.length > 0, "No user details provided");
+        // Check on the maximum size over which the for loop will run over.
+        require(currentAddresses.length <= 5, "Too many userss to whitelist");
+        for (uint256 i = 0; i < currentAddresses.length; i++) {
+            delete whitelist[currentAddresses[i]];
+			emit Whitelisted(currentAddresses[i],false);
```
Comment (dnzdlklc) [inline contracts/factoryContracts/SBTFactory.sol:171]: indendation

**Review by dnzdlklc — COMMENTED** [review]
See comments

File: contracts/factoryContracts/SBTFactory.sol (hunk 3 above)
Comment (atif-rehman-ideofuzion) [reply contracts/factoryContracts/SBTFactory.sol]: I think we don't need this event as its happening in issueOne and issueMany which already emits Issued event. Removing the event.

File: contracts/factoryContracts/SBTFactory.sol (hunk 6 above)
Comment (atif-rehman-ideofuzion) [reply contracts/factoryContracts/SBTFactory.sol:160]: don't see this issue in latest code as it indentation is already proper!y
image

File: contracts/factoryContracts/SBTFactory.sol [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+        _auth[id] = BurnAuth.Both;
+        nftToOwners[id] = _recipient;
+        _setUser(id, _recipient, _expires);
+
+        emit Issued(msg.sender, _recipient, id, _auth[id]);
+    }
+
+    /* @dev  MRHB or the whitelisted member can call this function to mint/issue 5 SBT at at time*/
+    function issueMany(
+        address[] memory _recipient,
```
Comment (dnzdlklc) [inline contracts/factoryContracts/SBTFactory.sol:83]: Revert to have tokenURI and mapping again.

**Review by dnzdlklc — COMMENTED** [review]
Revert uri changes and we're good to merge.

**Pull request merged**